ANGLE: Metal: Allow optimization of simple loops

kkinnunen-apple opened the pull request:
#### 1dbcd409197bfeda0417246083637b13f77e4f13
<pre>
ANGLE: Metal: Allow optimization of simple loops
<a href="https://bugs.webkit.org/show_bug.cgi?id=294109">https://bugs.webkit.org/show_bug.cgi?id=294109</a>
<a href="https://rdar.apple.com/141993099">rdar://141993099</a>

Reviewed by Mike Wyrzykowski.

Reimplement the feature to avoid undefined behavior of infinite loops.

Add EnsureLoopForwardProgress rewrite pass that inserts a volatile
variable access to all loops that it cannot analyze as being finite.

Detect loops of form `for (; i &lt;op&gt; x; ++i)` as being finite.
The &lt;op&gt; can be any of &lt;,&lt;=,&gt;,&gt;=,==, != operator.
The i can be int or uint.
The ++i can be -- or ++, -=1, +=1.
This assumes that backends using the feature emit signed int arithmetic
with defined wraparound semantics.

Uses volatile write instead of asm(&quot;&quot;) due to asm not forcing the
behavior in some compiler versions. The volatile variable access is
defined in C++ as forward progress, and by inheritance this works in
MSL.

Later commits may remove injectAsmStatementIntoLoopBodies if
ensureLoopForwardProgress is appropriate for all use-cases.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerSupport.h:
* Source/ThirdParty/ANGLE/include/GLSLANG/ShaderLang.h:
* Source/ThirdParty/ANGLE/include/platform/autogen/FeaturesMtl_autogen.h:
* Source/ThirdParty/ANGLE/include/platform/mtl_features.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/Static_builtins.json:
* Source/ThirdParty/ANGLE/src/compiler.gni:
* Source/ThirdParty/ANGLE/src/compiler/fuzz/translator_fuzzer.cpp:
(LLVMFuzzerTestOneInput):
* Source/ThirdParty/ANGLE/src/compiler/translator/Compiler.cpp:
(sh::TCompiler::checkAndSimplifyAST):
* Source/ThirdParty/ANGLE/src/compiler/translator/ImmutableString_autogen.cpp:
* Source/ThirdParty/ANGLE/src/compiler/translator/Operator_autogen.h:
* Source/ThirdParty/ANGLE/src/compiler/translator/SymbolTable_autogen.cpp:
(sh::Func::loopForwardProgress_):
(sh::TSymbolTable::findBuiltIn const):
(sh::TSymbolTable::isUnmangledBuiltInName const):
* Source/ThirdParty/ANGLE/src/compiler/translator/builtin_function_declarations.txt:
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/EmitMetal.cpp:
(GetOperatorString):
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/ProgramPrelude.cpp:
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/EnsureLoopForwardProgress.cpp: Added.
(sh::EnsureLoopForwardProgress):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/EnsureLoopForwardProgress.h: Added.
(sh::EnsureLoopForwardProgress):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_util/BuiltIn_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeFeatures):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ShaderMtl.mm:
(rx::ShaderMtl::compile):
* Source/ThirdParty/ANGLE/src/tests/angle_unittests.gni:
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/EnsureLoopForwardProgress_test.cpp: Added.
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/ImmutableString_test_autogen.cpp:
(sh::TEST(ImmutableStringTest, ScriptGeneratedHashesMatch)):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/MSLOutput_test.cpp:
((MSLOutputTest, EnsureLoopForwardProgressInfinite)):
((MSLOutputTest, EnsureLoopForwardProgressFinite)):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/Parse_test.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/GLSLTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TimeoutDrawTest.cpp:
(angle::TimeoutDrawTest::TimeoutDrawTest): Deleted.
(): Deleted.
(angle::TEST_P): Deleted.
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.cpp:
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):

Canonical link: <a href="https://commits.webkit.org/295932@main">https://commits.webkit.org/295932@main</a>
</pre>

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a079287a29ce4cfae1ae2f9040bcd7a0dfe103d2. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46416

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/fb41687ad6aab95b19c41742ee66b8c353b9f037. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46416

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/1b3ba2445f761b52614d6ede5e3743f38867e7bd. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/46416

kkinnunen-apple committed:
ANGLE: Metal: Allow optimization of simple loops
https://bugs.webkit.org/show_bug.cgi?id=294109
rdar://141993099

Reviewed by Mike Wyrzykowski.

Reimplement the feature to avoid undefined behavior of infinite loops.

Add EnsureLoopForwardProgress rewrite pass that inserts a volatile
variable access to all loops that it cannot analyze as being finite.

Detect loops of form `for (; i <op> x; ++i)` as being finite.
The <op> can be any of <,<=,>,>=,==, != operator.
The i can be int or uint.
The ++i can be -- or ++, -=1, +=1.
This assumes that backends using the feature emit signed int arithmetic
with defined wraparound semantics.

Uses volatile write instead of asm("") due to asm not forcing the
behavior in some compiler versions. The volatile variable access is
defined in C++ as forward progress, and by inheritance this works in
MSL.

Later commits may remove injectAsmStatementIntoLoopBodies if
ensureLoopForwardProgress is appropriate for all use-cases.

* Source/ThirdParty/ANGLE/ANGLE.xcodeproj/project.pbxproj:
* Source/ThirdParty/ANGLE/WebKit/TranslatorFuzzerSupport.h:
* Source/ThirdParty/ANGLE/include/GLSLANG/ShaderLang.h:
* Source/ThirdParty/ANGLE/include/platform/autogen/FeaturesMtl_autogen.h:
* Source/ThirdParty/ANGLE/include/platform/mtl_features.json:
* Source/ThirdParty/ANGLE/scripts/code_generation_hashes/Static_builtins.json:
* Source/ThirdParty/ANGLE/src/compiler.gni:
* Source/ThirdParty/ANGLE/src/compiler/fuzz/translator_fuzzer.cpp:
(LLVMFuzzerTestOneInput):
* Source/ThirdParty/ANGLE/src/compiler/translator/Compiler.cpp:
(sh::TCompiler::checkAndSimplifyAST):
* Source/ThirdParty/ANGLE/src/compiler/translator/ImmutableString_autogen.cpp:
* Source/ThirdParty/ANGLE/src/compiler/translator/Operator_autogen.h:
* Source/ThirdParty/ANGLE/src/compiler/translator/SymbolTable_autogen.cpp:
(sh::Func::loopForwardProgress_):
(sh::TSymbolTable::findBuiltIn const):
(sh::TSymbolTable::isUnmangledBuiltInName const):
* Source/ThirdParty/ANGLE/src/compiler/translator/builtin_function_declarations.txt:
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/EmitMetal.cpp:
(GetOperatorString):
* Source/ThirdParty/ANGLE/src/compiler/translator/msl/ProgramPrelude.cpp:
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/EnsureLoopForwardProgress.cpp: Added.
(sh::EnsureLoopForwardProgress):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_ops/msl/EnsureLoopForwardProgress.h: Added.
(sh::EnsureLoopForwardProgress):
* Source/ThirdParty/ANGLE/src/compiler/translator/tree_util/BuiltIn_autogen.h:
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/DisplayMtl.mm:
(rx::DisplayMtl::initializeFeatures):
* Source/ThirdParty/ANGLE/src/libANGLE/renderer/metal/ShaderMtl.mm:
(rx::ShaderMtl::compile):
* Source/ThirdParty/ANGLE/src/tests/angle_unittests.gni:
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/EnsureLoopForwardProgress_test.cpp: Added.
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/ImmutableString_test_autogen.cpp:
(sh::TEST(ImmutableStringTest, ScriptGeneratedHashesMatch)):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/MSLOutput_test.cpp:
((MSLOutputTest, EnsureLoopForwardProgressInfinite)):
((MSLOutputTest, EnsureLoopForwardProgressFinite)):
* Source/ThirdParty/ANGLE/src/tests/compiler_tests/Parse_test.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/GLSLTest.cpp:
* Source/ThirdParty/ANGLE/src/tests/gl_tests/TimeoutDrawTest.cpp:
(angle::TimeoutDrawTest::TimeoutDrawTest): Deleted.
(): Deleted.
(angle::TEST_P): Deleted.
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.cpp:
* Source/ThirdParty/ANGLE/util/autogen/angle_features_autogen.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::initializeEGLDisplay):

Canonical link: https://commits.webkit.org/295932@main
